src/game.cpp

```diff
@@ -6556,193 +6579,261 @@ void game::monmove()
+    sound( x, y, vol, description, true );
 }
 
-bool game::sound(int x, int y, int vol, std::string description, bool ambient)
+void game::sound(int x, int y, int vol, std::string description, bool ambient)
 {
-    // --- Monster sound handling here ---
-    // Alert all hordes
-    if (vol > 20 && levz == 0) {
-        int sig_power = ((vol > 140) ? 140 : vol) - 20;
-        cur_om->signal_hordes(levx + (MAPSIZE / 2), levy + (MAPSIZE / 2), sig_power);
+    if( vol == 0 ) {
+        // Bail out if no volume.
+        // TODO: log an error, this shouldn't happen?
+        return;
     }
-    // Alert all monsters (that can hear) to the sound.
-    for (int i = 0, numz = num_zombies(); i < numz; i++) {
-        monster &critter = critter_tracker.find(i);
-        // rl_dist() is faster than critter.has_flag() or critter.can_hear(), so we'll check it first.
-        int dist = rl_dist(x, y, critter.posx(), critter.posy());
-        int vol_goodhearing = vol * 2 - dist;
-        if (vol_goodhearing > 0 && critter.can_hear()) {
-            const bool goodhearing = critter.has_flag(MF_GOODHEARING);
-            int volume = goodhearing ? vol_goodhearing : (vol - dist);
-            // Error is based on volume, louder sound = less error
-            if (volume > 0) {
-                int max_error = 0;
-                if (volume < 2) {
-                    max_error = 10;
-                } else if (volume < 5) {
-                    max_error = 5;
-                } else if (volume < 10) {
-                    max_error = 3;
-                } else if (volume < 20) {
-                    max_error = 1;
-                }
+    recent_sounds.emplace_back( std::make_pair( point(x, y), vol ) );
+    sounds_since_last_turn.emplace(
+        std::make_pair( point(x, y), sound_event{vol, description, ambient, false} ) );
+}
+
+void game::add_footstep(int x, int y, int volume, int, monster *)
+{
+    sounds_since_last_turn.emplace(
+        std::make_pair( point(x, y), sound_event{volume, "", false, true} ) );
+}
+
+template <typename C>
+static void vector_quick_remove( std::vector<C> &source, int index )
+{
+    if( source.size() != 1 ) {
+        // Swap the target and the last element of the vector.
+        // This scrambles the vector, but makes removal O(1).
+        std::iter_swap( source.begin() + index, source.end() - 1 );
+    }
+    source.pop_back();
+}
+
+static std::vector<centroid> cluster_sounds( std::vector<std::pair<point, int>> recent_sounds )
+{
+    // If there are too many monsters and too many noise sources (which can be monsters, go figure),
+    // applying sound events to monsters can dominate processing time for the whole game,
+    // so we cluster sounds and apply the centroids of the sounds to the monster AI
+    // to fight the combanatorial explosion.
+    std::vector<centroid> sound_clusters;
+    const int num_seed_clusters = std::max( std::min(recent_sounds.size(), (size_t)10),
```

Comment:
This is what decides how many clusters to use.  In practice it's almost always 10 or less because log(num_sounds) is always less than 10.  We can pretty easily handle larger numbers, but if they get much larger it's problematic, because the work scales with (num_sound_events * num_monsters)